### ISSUE TITLE
Efficient read of subcolumns from tables.

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#14196]
**Use case**

Scenario 1:
User has column of type Tuple in table, example:
```
CREATE TABLE t (x Tuple(Int8, String)) ENGINE = MergeTree ...
```
If they write 
```
SELECT x.2 FROM t
```
now both subcolumns of tuple will be read (we will read whole `x` column).
We want to read only required subcolumn.

Scenario 2 (done, #17310):
Users can create a table with Nested fields.
They already can query subcolumns:
```
SELECT nested.x, nested.y FROM table
```
But it is implemented in very special way.
And users cannot query all nested column as an array of tuples:
```
SELECT nested FROM table
```
Neither they can do arrayJoin on nested as a whole:
```
SELECT arrayJoin(nested) FROM table
```
We want to rectify the code around support for Nested fields.

Scenario 3 (done, #17310):
Allow to have Nested fields with arbitrary nesting:
```
CREATE TABLE t (x Nested(a Int8, b Nested(u String, v Float32))) ENGINE = MergeTree ...
SELECT x.a, x.b.u, x.b.v FROM t;
```

Scenario 4:
Allow to write `nested.*` expressions.

Scenario 5 (done, #24406):
When user is only using array length in query, we don't have to read array contents - we can only read subcolumn with length:
```
SELECT length(arr) FROM table
```
Can be rewritten to
```
SELECT arr.size0 FROM table
```
And size0 is the name of subcolumn that we already have on disk.

Scenario 6 (done, #17310):
We already support named tuples.
Allow to write `x.name` for named tuples. Now only the syntax `tupleElement(x, 'name')` is possible.

Scenario 7:
Support the case when the existence and the types of subcolumns are determined at query time, not at table creation time.
It can be represented by special data type for semistructured data.

**Comment by alexey-milovidov** [comment]
We already have half of the fundamentals to implement this task.

How to get started implementing it:
- we already restrict using of dots in column names;
- pass column names with dots as is to IStorage::read;
- blocks/chunks will have names with dots passed around as we already do for Nested;
- edit the `StorageInMemoryMetadata`: `check` and `getSampleBlockForColumns` methods to allow subcolumns;
- rewrite `IDataType::enumerateStreams` to allow extraction of nested data types; maybe add another method to obtain data type for subpath.
...

**Comment by CurtizJ** [comment]
> we already restrict using of dots in column names

But actually we can use dots and everything by wrapping column name in backticks. BTW, it doesn't look like a big problem, because we can match column name from query with column names in storage definition, cut its prefix and correctly interpret subcolumn name.

**Comment by CurtizJ** [comment]
I will write here corner cases and caveats, that I found.

**Comment by CurtizJ** [comment]
> we can match column name from query with column names in storage definition, cut its prefix and correctly interpret subcolumn name

There are some corner cases with ambiguous column names. E.g. we can create table with following structure:

```
CREATE TABLE t (`s.1` String, s Tuple(Int, Int)) ENGINE = MergeTree ORDER BY tuple();
```

But it will lead to undefined behaviour (in terms of ClickHouse)
```
INSERT INTO t VALUES ('foo', (1, 2));

SELECT * FROM t;

Received exception from server (version 20.10.1):
Code: 246. DB::Exception: Received from localhost:9000. DB::Exception: Bad size of marks file './data/default/t/all_1_1_0/s%2E1.mrk2': 96, must be: 48: While executing MergeTreeThread. 

SELECT s.1 FROM t

Received exception from server (version 20.10.1):
Code: 246. DB::Exception: Received from localhost:9000. DB::Exception: Bad size of marks file './data/default/t/all_1_1_0/s%2E1.mrk2': 96, must be: 48: While executing MergeTreeThread.
```

Since it doesn't work now we can restrict it. But some exotic cases of new features should be also restricted, like the following nested of nested, which is ok I think.

```
(`n.b` Nested(u UInt, s String), n Nested(x UInt8, b Nested(u UInt32, s String)))

**Comment by CurtizJ** [comment]
Several cases were implemented in #17310.

**Comment by geertvanbommel** [comment]
I have a curious question.  In the June meetup this example was given by Alexey:
```sql
CREATE TABLE test (id UInt64, data Object('JSON'))
ENGINE = MergeTree;

INSERT INTO test FORMAT JSONEachRow
{"id": 1, "data": {"kl": "aa", "k2": {"k3": "bb", "64": 123}}}
{"id": 2, "data": {"kl": "ee", "k5": "ff"}};

SELECT id, data.ki, data.k2.*, data.k2.k4::String FROM test;
```
Would querying `data.*.k4` also be possible? 
```sql
SELECT id, data.ki, data.k2.*, data.*.k4::String FROM test;
```

And also, what if `data.k2` was an array? In other words, would nested arrays be supported?
